### PR TITLE
Implement backing layer in metal backend, [MAGMA-3664]

### PR DESCRIPTION
**Pull request by hsong-gpsw**
1. Implement get_backing_layer() in metal for macOS and iOS
2. Support passing null as window and getting generated layer from get_backing_layer() in ngl-tools for macOS
3. Add get_backing_layer() binding in pynodegl python.

### CODE REVIEW

File: libnodegl/src/backends/metal/api_mtl.m [hunk 1]
```diff
@@ -30,11 +30,12 @@
 #include <QuartzCore/QuartzCore.h>
 
 #include "log.h"
+#include "gpu_ctx_mtl_c.h"
```
Comment (jstys-gpsw) [inline libnodegl/src/backends/metal/api_mtl.m:33]: Where does this include come from?

File: libnodegl/src/backends/metal/gpu_ctx_mtl.cpp [hunk 2]
```diff
@@ -25,6 +25,7 @@
 #define MTL_PRIVATE_IMPLEMENTATION
 
 #include "gpu_ctx_mtl.h"
+#include "gpu_ctx_mtl_c.h"
```
Comment (jstys-gpsw) [inline libnodegl/src/backends/metal/gpu_ctx_mtl.cpp:28]: Same include here

File: libnodegl/src/backends/metal/api_mtl.m (hunk 1 above)
Comment (hsong-gpsw) [reply libnodegl/src/backends/metal/api_mtl.m:33]: I forget to add this new file.  The purpose of this new header is to get function prototype "int ngl_gpu_ctx_mtl_get_backing_layer()". Added

File: libnodegl/src/backends/metal/gpu_ctx_mtl.cpp (hunk 2 above)
Comment (hsong-gpsw) [reply libnodegl/src/backends/metal/gpu_ctx_mtl.cpp:28]: I forget to add this new file. The purpose of this new header is to get function prototype "int ngl_gpu_ctx_mtl_get_backing_layer()". Added

File: ngl-tools/wsi_windows.c [hunk 3]
```diff
@@ -43,3 +43,9 @@ int wsi_set_ngl_config(struct ngl_config *config, SDL_Window *window)
 
     return -1;
 }
+
+int wsi_set_backing_layer(struct ngl_config *config, SDL_Window *window,
+                          void *backing_layer)
+{
+    return 0;
```
Comment (jstys-gpsw) [inline ngl-tools/wsi_windows.c:50]: Should this return -1 actually?

File: ngl-tools/wsi_linux.c [hunk 4]
```diff
@@ -52,3 +52,9 @@ int wsi_set_ngl_config(struct ngl_config *config, SDL_Window *window)
 
     return -1;
 }
+
+int wsi_set_backing_layer(struct ngl_config *config, SDL_Window *window,
+                          void *backing_layer)
+{
+    return 0;
```
Comment (jstys-gpsw) [inline ngl-tools/wsi_linux.c:59]: return -1 here as well

File: ngl-tools/wsi_cocoa.m [hunk 5]
```diff
@@ -86,3 +89,33 @@ int wsi_destroy_ngl_config(struct ngl_config *config) {
...
+    SDL_VERSION(&info.version);
+    if (!SDL_GetWindowWMInfo(window, &info)) {
+        fprintf(stderr, "Failed to get window WM information\n");
+        return -1;
+    }
+
+    NSWindow *nswindow = info.info.cocoa.window;
+    NSView *view = [nswindow contentView];
+    if (info.subsystem == SDL_SYSWM_COCOA) {
+        if (config->backend == NGL_BACKEND_VULKAN ||
```
Comment (jstys-gpsw) [inline ngl-tools/wsi_cocoa.m:111]: Should `NGL_BACKEND_VULKAN` be dropped here?

File: tests/api.py [hunk 6]
```diff
@@ -90,8 +90,13 @@ def api_reconfigure_fail():
     assert ctx.set_scene(scene) == 0
     assert ctx.draw(0) == 0
     ret = ctx.configure(offscreen=0, backend=_backend)
-    assert ret != 0
-    assert ctx.draw(1) != 0
+    backing_layer = ctx.get_backing_layer()
```
[the remark this answers is not in the document]
Comment (hsong-gpsw) [reply tests/api.py:93]: changed.  Thanks for suggestion

File: ngl-tools/wsi_linux.c (hunk 4 above)
Comment (hsong-gpsw) [reply ngl-tools/wsi_linux.c:59]: Yes, changed.  Thanks

File: ngl-tools/wsi_cocoa.m (hunk 5 above)
Comment (hsong-gpsw) [reply ngl-tools/wsi_cocoa.m:111]: Right.  Changed.  Thanks for review

**Review by jstys-gpsw — APPROVED** [review]

**Pull request merged**